### ISSUE TITLE
Better caching

### ISSUE THREAD

**Issue by OhadRubin** [allenai/allennlp#4591]
Something along the lines of what fairseq has:
[Here](https://github.com/pytorch/fairseq/blob/master/fairseq/data/indexed_dataset.py
)
For example, in MMapIndexedDataset, they calculate the sizes for each field and then read exactly that size into an array.
[Here](https://github.com/allenai/allennlp/pull/4497#issuecomment-667314473), @matt-gardner mentions that this is the type of caching that you guys might implement. A database that just stores tensors directly.

**Comment by matt-gardner** [comment]
There is some initial work along these lines [here](https://github.com/allenai/allennlp/blob/0a4765c4e978374a6665139ccf9f235f833c0e7f/allennlp/common/file_utils.py#L333-L404) (used in a dataset reader [here](https://github.com/allenai/allennlp/blob/0a4765c4e978374a6665139ccf9f235f833c0e7f/allennlp/data/dataset_readers/nlvr2.py#L114-L116)).  I think we definitely want to figure out how to include this in our main `Instance` creation pipeline, but that'll come a little bit later.  If you're interested in helping to design / work on this, contributions welcome.  This is a big enough piece of work that some kind of design document (or detailed github description) should come before any code is written.

@epwalsh, do we have an issue open already for this?  If not, I'll keep this open and add it to either the 2.0 or the performance milestone (maybe 2.0?).

**Comment by epwalsh** [comment]
@matt-gardner I don't think we had a separate issue tracking caching until now, so adding this to 2.0 sounds good to me.

**Comment by OhadRubin** [comment]
Should I create a .md page or a Google Docs? 
Anyway, most of the logic will use the to_tensor method of each field, right?
What I was thinking was:
For each instance, call to_tensor on all of its fields and then either:
1. use TensorCache.
2. call .numpy() and record the sizes, flatten the data and use logic from MMapIndexedDataset.
For fields that don't have to_tensor, like MetadataField, use jsonpickle.

**Comment by epwalsh** [comment]
Hey @OhadRubin, I've already started a design document which I'll link to on this thread once it's ready to be seen.

**Comment by epwalsh** [comment]
There is a good discussion here about `lmdb`: https://github.com/allenai/allennlp/pull/4578#issuecomment-680776842.

I'm working on an API design that would be agnostic to the backend or ser/deserialization method we use, so we can decide on that later.

**Comment by OhadRubin** [comment]
So it seems there are different use cases for caching, maybe I can write some code that extends `TensorCache`  into a `Registrable`  and implement a `MMapIndexedDataset` style `TensorCache`.
@epwalsh, what do you think?

**Comment by epwalsh** [comment]
@OhadRubin that would be great if you got started on that!

Right now just I'm focusing on the overall API, and how that would integrate into our data pipeline, so I don't think that collides with what you want to work on. Once I'm finished with the API skeleton we should be able to plug in the `TensorCache` / `MMapIndexedDataset` / whatever we go with.

**Comment by OhadRubin** [comment]
So I should inherit from DatasetReader and override _instances_to_cache_file, correct?
I understand there are plans to move caching to DataLoader, but for now, that should be ok, right?

**Comment by epwalsh** [comment]
@OhadRubin actually no, we're working off of the `vision` branch now, which will become AllenNLP 2.0. In that branch the caching mechanism in the `DatasetReader` has been removed. Maybe just start by implementing this in a standalone file?

**Comment by OhadRubin** [comment]
Ok, so I'll assume I am iterating over objects with a serialize method, is that ok?

**Comment by epwalsh** [comment]
Yes, but we can assume these objects are indexed `Instance`s.

**Comment by OhadRubin** [comment]
Hey, @epwalsh, after looking a bit more into `MMapIndexedDataset`, much of the benefit of it comes from knowing the structure and sizes of everything, can I assume `serialize` returns a `Dict[str, Union[torch.Tensor, str]`?

**Comment by epwalsh** [comment]
Hey @OhadRubin, well after the talking with the team a little more yesterday we decided it would probably be more beneficial to cache tensor dicts instead of actual `Instance` objects. In other words, we want to be caching the result of [`Instance.as_tensor_dict()`](https://github.com/allenai/allennlp/blob/master/allennlp/data/instance.py#L87-L89).

So in most cases, this is a `Dict[str, torch.Tensor]`, which would play nicely with the `MMapIndexedDataset` approach. However, there are some exceptions. In particular, when an `Instance` contains a `MetaDataField`, `Instance.as_dict_dict()` could contain pretty much anything.

**Comment by OhadRubin** [comment]
I understand.

**Comment by OhadRubin** [comment]
There is a overhead  of around 760 bytes for saving a `Tensor` with `torch.save`, compared to numpy.tobytes.
So i'll use numpy.
```
import torch
import io
import pickle
import numpy as np

for dtype in [torch.bool, torch.float16,torch.float32,torch.int16,torch.int32,torch.int64]:
    print(dtype)
    for size in range(2,5):
        in_list = list(range(10**size))
        torch_tensor = torch.tensor(in_list,dtype=dtype).detach()
        buffer = io.BytesIO()
        torch.save(torch_tensor, buffer, pickle_protocol=pickle.HIGHEST_PROTOCOL)
        torch_res = buffer.getbuffer().tobytes()
        np_res = torch_tensor.numpy().tobytes(order='C')
        print(f"size: {10**size} - torch_res: {len(torch_res)}, np_res: {len(np_res)}")
```
Output:
```
torch.bool
size: 100 - torch_res: 824, np_res: 100
size: 1000 - torch_res: 1720, np_res: 1000
size: 10000 - torch_res: 10744, np_res: 10000
torch.float16
size: 100 - torch_res: 952, np_res: 200
size: 1000 - torch_res: 2744, np_res: 2000
size: 10000 - torch_res: 20728, np_res: 20000
torch.float32
size: 100 - torch_res: 1144, np_res: 400
size: 1000 - torch_res: 4728, np_res: 4000
size: 10000 - torch_res: 40760, np_res: 40000
torch.int16
size: 100 - torch_res: 952, np_res: 200
size: 1000 - torch_res: 2744, np_res: 2000
size: 10000 - torch_res: 20728, np_res: 20000
torch.int32
size: 100 - torch_res: 1144, np_res: 400
size: 1000 - torch_res: 4728, np_res: 4000
size: 10000 - torch_res: 40760, np_res: 40000
torch.int64
size: 100 - torch_res: 1528, np_res: 800
size: 1000 - torch_res: 8760, np_res: 8000
size: 10000 - torch_res: 80760, np_res: 80000
```

**Comment by epwalsh** [comment]
API design document (work in progress): https://docs.google.com/document/d/1EBAcPF19NM7bYuwDKmHN4Ws361p0Eo4vjIpSA67l1rQ/edit?usp=sharing

**Comment by dirkgr** [comment]
`numpy.tobytes()` does not remember the data type, size, or value order of the tensor. There might be other stuff. `torch.save()` does, and we need that stuff to be preserved.

I would consider using `numpy.tobytes()` and `torch.from_numpy()` anyways because it allows us to create a torch tensor that's backed by memory-mapped bytes, no reading required. But we have some other hurdles to clear before we can do that. So for now, let's stick with `torch.save()`.

**Comment by dirkgr** [comment]
Since reading is so much faster now, the case for caching is pretty weak. We're abandoning this issue for now.
